keras/src/layers/preprocessing/image_preprocessing/random_grayscale.py

```diff
@@ -0,0 +1,120 @@
+    """Preprocessing layer for random conversion of RGB images to grayscale.
+
+    This layer randomly converts input images to grayscale with a specified
+    probability. When applied, it maintains the original number of channels
+    but sets all channels to the same grayscale value. This can be useful
+    for data augmentation and training models to be robust to color
+    variations.
+
+    The conversion preserves the perceived luminance of the original color
+    image using standard RGB to grayscale conversion coefficients. Images
+    that are not selected for conversion remain unchanged.
+
+    **Note:** This layer is safe to use inside a `tf.data` pipeline
+    (independently of which backend you're using).
+
+    Args:
+        probability: Float between 0 and 1, specifying the probability of
+            converting each image to grayscale. Defaults to 0.5. A value of
+            1.0 means all images will be converted, while 0.0 means no images
+            will be converted.
+        data_format: String, one of `"channels_last"` (default) or
+            `"channels_first"`. The ordering of the dimensions in the inputs.
+            `"channels_last"` corresponds to inputs with shape
+            `(batch, height, width, channels)` while `"channels_first"`
+            corresponds to inputs with shape
+            `(batch, channels, height, width)`.
+
+    Input shape:
+        3D (unbatched) or 4D (batched) tensor with shape:
+        `(..., height, width, channels)`, in `"channels_last"` format,
+        or `(..., channels, height, width)`, in `"channels_first"` format.
+
+    Output shape:
+        Same as input shape. The output maintains the same number of channels
+        as the input, even for grayscale-converted images where all channels
+        will have the same value.
+
+    Example:
+
+    ```python
+    # Create a random grayscale layer with 50% probability
+    random_gray = keras.layers.RandomGrayscale(probability=0.5)
+
+    # Apply to a single color image
+    color_image = [...] # your input RGB image
+    output_image = random_gray(color_image)
+
+    # Use in a sequential model for data augmentation
+    model = keras.Sequential([
+        keras.layers.RandomGrayscale(probability=0.3),
+        keras.layers.Conv2D(16, 3, activation='relu'),
+        # ... rest of your model
+    ])
+    ```
+    """
+
+    def __init__(self, probability=0.5, data_format=None, **kwargs):
+        super().__init__(**kwargs)
+        if probability < 0 or probability > 1:
+            raise ValueError(
+                "`probability` should be between 0 and 1. "
+                f"Received: probability={probability}"
+            )
+        self.probability = probability
+        self.data_format = backend.standardize_data_format(data_format)
+        self.random_generator = self.backend.random.SeedGenerator()
+
+    def transform_images(self, images, transformations=None, **kwargs):
```

Comment:
Let's stick to the usual factoring, where the transformation is defined in a separate function